lib/python/qmk/cli/c2json.py

```diff
@@ -0,0 +1,324 @@
+from enum import auto, Flag
+import itertools
+import json
+import os
+from pathlib import Path
+from typing import List, NamedTuple
+
+from milc import cli
+import pycparser
+from pycparser.c_ast import BinaryOp, Constant, Decl, ID
+from pycparserext.ext_c_parser import GnuCParser
+
+
+def iter_upwards(path):
+    """
+    Generator to iterate through all parent directories yielding a Path
+    for each
+
+    Parameters:
+    path -- Path object to walk through
+    """
+    parts = list(path.resolve().parts)
+    while parts:
+        yield Path(*parts)
+        parts.pop()
+
+
+def find_topmost_keyboard_h(path):
+    """
+    Look through all directories up from the given path to find a keyboard.h
+    file. This is used for resolving QMK_KEYBOARD_H.
+
+    Parameters:
+    path -- path to the keymap.h we are resolving for
+
+    Returns:
+    A Path object for the keyboard.h found or None if we can't find one
+
+    Signature:
+    (path: Path) -> Path
+    """
+
+    def header_name(directory: Path) -> str:
+        return directory.name + '.h'
+
+    # path will look like
+    # keyboards/NAME/SUBNAME/keymaps/keymap.c
+    keyboard_h = None
+    for part in iter_upwards(path):
+        # if we hit a root level subdirectory we've overshot it
+        if part.name in ('keyboards', 'users', 'layouts'):
+            return keyboard_h
+
+        header = part / header_name(part)
+        if header.exists():
+            keyboard_h = header
+
+
+def get_qmk_root(from_path):
+    """
+    Finds the path of the QMK repository root relative to a keymaps directory
+
+    Parameters:
+    from_path -- Path to find the QMK root relative to
+
+    Signature:
+    (from_path: Path) -> Path
+    """
+    for part in iter_upwards(from_path.absolute()):
+        if (part / 'keyboards').exists():
+            return part
```

Comment:
You can always assume that os.getcwd() is the qmk_firmware root, you shouldn't ever need to look for it.